File: libtao/src/sharedarrays.c
/*
 * sharedarrays.c --
 *
 * Implement multi-dimensional arrays whose contents can be shared between
 * processes.
 *
 *-----------------------------------------------------------------------------
 *
 * This file if part of the TAO library (https://github.com/emmt/TAO) licensed
 * under the MIT license.
 *
 * Copyright (C) 2018-2019, Éric Thiébaut.
 */

#include "config.h"
#include "macros.h"
#include "tao-private.h"

int
tao_get_shared_array_ident(const tao_shared_array_t* arr)
{
    return (TAO_LIKELY(arr != NULL) ? arr->base.ident : -1);
}

tao_element_type_t
tao_get_shared_array_eltype(const tao_shared_array_t* arr)
{
    return (TAO_LIKELY(arr != NULL) ? arr->eltype : -1);
}

long
tao_get_shared_array_length(const tao_shared_array_t* arr)
{
    return (TAO_LIKELY(arr != NULL) ? arr->nelem : 0);
}

int
tao_get_shared_array_ndims(const tao_shared_array_t* arr)
{
    return (TAO_LIKELY(arr != NULL) ? arr->ndims : 0);
}

long
tao_get_shared_array_size(const tao_shared_array_t* arr, int d)
{
    return (TAO_LIKELY(arr != NULL) ?
            (TAO_UNLIKELY(d < 1) ? 0 :
             (TAO_UNLIKELY(d > TAO_MAX_NDIMS) ? 1 :
              arr->dims[d-1])) : 0);
}

void*
tao_get_shared_array_data(const tao_shared_array_t* arr)
{
    return (TAO_LIKELY(arr != NULL) ?
            (void*)((uint8_t*)arr + arr->offset) : (void*)0);
}

int
tao_get_shared_array_nreaders(const tao_shared_array_t* arr)
{
    return (TAO_LIKELY(arr != NULL) ? arr->nreaders : 0);
}

int
tao_adjust_shared_array_nreaders(tao_shared_array_t* arr, int adj)
{
    return (TAO_LIKELY(arr != NULL) ? (arr->nreaders += adj) : 0);
}

int
tao_get_shared_array_nwriters(const tao_shared_array_t* arr)
{
    return (TAO_LIKELY(arr != NULL) ? arr->nwriters : 0);
}

int
tao_adjust_shared_array_nwriters(tao_shared_array_t* arr, int adj)
{
    return (TAO_LIKELY(arr != NULL) ? (arr->nwriters += adj) : 0);
}

int64_t
tao_get_shared_array_counter(const tao_shared_array_t* arr)
{
    return (TAO_LIKELY(arr != NULL) ? arr->counter : -1);
}

void
tao_set_shared_array_counter(tao_shared_array_t* arr, int64_t cnt)
{
    if (TAO_LIKELY(arr != NULL)) {
        arr->counter = cnt;
    }
}

void
tao_get_shared_array_timestamp(const tao_shared_array_t* arr,
                               int64_t* ts_sec, int64_t* ts_nsec)
{
    if (TAO_LIKELY(arr != NULL)) {
        *ts_sec = arr->ts_sec;
        *ts_nsec = arr->ts_nsec;
    } else {
        *ts_sec = -1;
        *ts_nsec = 0;
    }
}

void
tao_set_shared_array_timestamp(tao_shared_array_t* arr,
                               int64_t ts_sec, int64_t ts_nsec)
{
    if (TAO_LIKELY(arr != NULL)) {
        arr->ts_sec = ts_sec;
        arr->ts_nsec = ts_nsec;
    }
}

tao_shared_array_t*
tao_create_1d_shared_array(tao_error_t** errs, tao_element_type_t eltype,
                           long dim1, unsigned int perms)
{
    long dims[1];
    dims[0] = dim1;
    return tao_create_shared_array(errs, eltype, 1, dims, perms);
}

tao_shared_array_t*
tao_create_2d_shared_array(tao_error_t** errs, tao_element_type_t eltype,
                           long dim1, long dim2, unsigned int perms)
{
    long dims[2];
    dims[0] = dim1;
    dims[1] = dim2;
    return tao_create_shared_array(errs, eltype, 2, dims, perms);
}

tao_shared_array_t*
tao_create_3d_shared_array(tao_error_t** errs, tao_element_type_t eltype,
                           long dim1, long dim2, long dim3,
                           unsigned int perms)
{
    long dims[3];
    dims[0] = dim1;
    dims[1] = dim2;
    dims[2] = dim3;
    return tao_create_shared_array(errs, eltype, 3, dims, perms);
}

tao_shared_array_t*
tao_create_shared_array(tao_error_t** errs, tao_element_type_t eltype,
                        int ndims, const long dims[], unsigned int perms)
{
    size_t elsize = tao_get_element_size(eltype);
    if (elsize < 1) {
        tao_push_error(errs, __func__, TAO_BAD_TYPE);
        return NULL;
    }
    long nelem = tao_count_elements(errs, ndims, dims);
    if (nelem < 1) {
        return NULL;
    }
    size_t offset = TAO_ROUND_UP(sizeof(tao_shared_array_t), ALIGNMENT);
    size_t nbytes = offset + nelem*elsize;
    tao_shared_object_t* obj = tao_create_shared_object(errs, TAO_SHARED_ARRAY,
                                                        nbytes, perms);
    if (obj == NULL) {
        return NULL;
    }
    tao_shared_array_t* arr = (tao_shared_array_t*)obj;
    arr->offset = offset;
    arr->nelem = nelem;
    arr->ndims = ndims;
    for (int d = 0; d < ndims; ++d) {
        arr->dims[d] = dims[d];
    }
    for (int d = ndims; d < TAO_MAX_NDIMS; ++d) {
        arr->dims[d] = 1;
    }
    arr->eltype = eltype;
    arr->nwriters = 0;
    arr->nreaders = 0;
    arr->counter = 0;
    struct timespec ts;
    if (clock_gettime(CLOCK_MONOTONIC, &ts) == -1) {
        /* Just ignore this error. */
        ts.tv_sec = 0;
        ts.tv_nsec = 0;
    }
    arr->ts_sec = ts.tv_sec;
    arr->ts_nsec = ts.tv_nsec;
    return arr;
}

tao_shared_array_t*
tao_attach_shared_array(tao_error_t** errs, int ident)
{
    return (tao_shared_array_t*)tao_attach_shared_object(errs, ident,
                                                         TAO_SHARED_ARRAY);
}

int
tao_detach_shared_array(tao_error_t** errs, tao_shared_array_t* arr)
{
    /* No needs to check address. */
    return tao_detach_shared_object(errs, (tao_shared_object_t*)arr);
}

int
tao_lock_shared_array(tao_error_t** errs, tao_shared_array_t* arr)
{
    if (TAO_UNLIKELY(arr == NULL)) {
        tao_push_error(errs, __func__, TAO_BAD_ADDRESS);
        return -1;
    }
    return tao_lock_mutex(errs, &arr->base.mutex);
}

int
tao_try_lock_shared_array(tao_error_t** errs, tao_shared_array_t* arr)
{
    if (TAO_UNLIKELY(arr == NULL)) {
        tao_push_error(errs, __func__, TAO_BAD_ADDRESS);
        return -1;
    }
    return tao_try_lock_mutex(errs, &arr->base.mutex);
}

int
tao_unlock_shared_array(tao_error_t** errs, tao_shared_array_t* arr)
{
    if (TAO_UNLIKELY(arr == NULL)) {
        tao_push_error(errs, __func__, TAO_BAD_ADDRESS);
        return -1;
    }
    return tao_unlock_mutex(errs, &arr->base.mutex);
}

File: julia/src/tao/libtao/sharedarrays.jl
#
# sharedarrays.jl --
#
# Management of shared multi-dimensional arrays for Julia interface to TAO
# C-library.
#
#-------------------------------------------------------------------------------
#
# This file if part of the TAO software (https://github.com/emmt/TAO) licensed
# under the MIT license.
#
# Copyright (C) 2018, Éric Thiébaut.
#

# Make a `TAO.SharedArray{T,N}` behaves like an array.  Note that, for
# performance reasons, reading/writing the array contents is done without any
# attempt to lock the object before.  Locking has to be done appropriately
# when such an object is used.
Base.eltype(obj::SharedArray{T,N}) where {T,N} = T
Base.length(obj::SharedArray) = length(obj.arr)
Base.ndims(obj::SharedArray{T,N}) where {T,N} = N
Base.size(obj::SharedArray) = size(obj.arr)
Base.size(obj::SharedArray, d) = size(obj.arr, d)
Base.axes(obj::SharedArray) = axes(obj.arr)
Base.axes(obj::SharedArray, d) = axes(obj.arr, d)
Base.eachindex(obj::SharedArray) = eachindex(obj.arr)
Base.stride(obj::SharedArray, d) = stride(obj.arr, d)
Base.strides(obj::SharedArray) = strides(obj.arr)
Base.firstindex(obj::SharedArray) = 1
@inline Base.firstindex(obj::SharedArray{T,N}, d) where {T,N} =
    ((d % UInt) - 1 < N ? 1 : error("dimension out of range"))
Base.lastindex(obj::SharedArray) = length(obj.arr)
Base.lastindex(obj::SharedArray, d) = size(obj.arr, d)

# FIXME: similar
Base.similar(obj::SharedArray, args...) = similar(obj.arr, args...)
Base.reshape(obj::SharedArray, dims...) = reshape(obj.arr, dims...)
Base.copy(obj::SharedArray) = copy(obj.arr)
Base.deepcopy(obj::SharedArray) = deepcopy(obj.arr)
Base.fill!(obj::SharedArray, val) = fill!(obj.arr, val)

@inline @propagate_inbounds Base.getindex(obj::SharedArray, inds...) =
    getindex(obj.arr, inds...)
@inline @propagate_inbounds Base.setindex!(obj::SharedArray, val, inds...) =
    setindex!(obj.arr, val, inds...)

Base.IndexStyle(::Type{<:SharedArray}) = IndexLinear()

# Make a `TAO.SharedArray{T,N}` iterable.
Base.iterate(obj::SharedArray) = iterate(obj.arr)
Base.iterate(obj::SharedArray, state) = iterate(obj.arr, state)

function create(::Type{SharedArray{T}}, dims::Integer...; kwds...) where {T}
    N = length(dims)
    return create(SharedArray{T,N}, convert(NTuple{N,Int}, dims); kwds...)
end

function create(::Type{SharedArray{T,N}},
                dims::NTuple{N,Int};
                perms::Integer = 0o600) where {T,N}
    @assert haskey(SHARED_ARRAY_ELTYPES, T) "Invalid array element type"
    eltype = SHARED_ARRAY_ELTYPES[T]
    errs = Errors()
    ptr = ccall((:tao_create_shared_array, taolib), Ptr{Cvoid},
                (Ref{Errors}, Cint, Cint, Ptr{Clong}, Cuint),
                errs, eltype, N, Clong[dims...], perms)
    _check(ptr != C_NULL, errs)
    return _wrap(SharedArray{T,N}, ptr, dims)
end

function attach(::Type{SharedArray}, ident::Integer)
    errs = Errors()
    ptr = ccall((:tao_attach_shared_array, taolib), Ptr{Cvoid},
                (Ref{Errors}, Cint), errs, ident)
    _check(ptr != C_NULL, errs)
    eltype = ccall((:tao_get_shared_array_eltype, taolib), Cint,
                   (Ptr{Cvoid},), ptr)
    1 ≤ eltype ≤ length(SHARED_ARRAY_TYPES) ||
        _detach_on_error(ptr, "Bad element type")
    N = convert(Int, ccall((:tao_get_shared_array_ndims, taolib), Cint,
                           (Ptr{Cvoid},), ptr))
    1 ≤ N ≤ SHARED_ARRAY_MAX_NDIMS ||
        _detach_on_error(ptr, "Bad number of dimensions")
    dims = Array{Int}(undef, N)
    for d in 1:N
        dims[d] = ccall((:tao_get_shared_array_size, taolib), Clong,
                        (Ptr{Cvoid}, Cint), ptr, d)
        dims[d] ≥ 1 || _detach_on_error(ptr, "Bad dimension $d")
    end
    T = SHARED_ARRAY_TYPES[eltype]
    return _wrap(SharedArray{T,N}, ptr, (dims...,))
end

function detach(obj::SharedArray{T,N}) where {T,N}
    if obj.ptr != C_NULL
        # Detaching a TAO shared array makes the associated Julia array invalid
        # so we replace it with an array of the correct type but with all
        # dimensions equal to zero.
        ptr = obj.ptr
        obj.arr = Array{T,N}(undef, ntuple(i -> 0, N))
        obj.ptr = C_NULL
        _check(_detach(ptr))
    end
end

function _detach_on_error(ptr::Ptr{Cvoid}, msg::AbstractString)
    errs = detach(ptr::Ptr{Cvoid})
    any_errors(errs) && discard_errors(errs)
    error(msg)
end

function _wrap(::Type{SharedArray{T,N}}, ptr::Ptr{Cvoid},
               dims::NTuple{N,Int}) where {T,N}
    data = ccall((:tao_get_shared_array_data, taolib), Ptr{T},
                 (Ptr{Cvoid},), ptr)
    arr = unsafe_wrap(Array, data, dims; own = false)
    obj = SharedArray{T,N}(ptr, arr)
    return finalizer(_destroy, obj)
end

get_nreaders(obj::SharedArray) =
    ccall((:tao_get_shared_array_nreaders, taolib), Cint,
          (Ptr{Cvoid},), obj)

adjust_nreaders!(obj::SharedArray, adj::Integer) =
    ccall((:tao_adjust_shared_array_nreaders, taolib), Cint,
          (Ptr{Cvoid}, Cint), obj, adj)

get_nwriters(obj::SharedArray) =
    ccall((:tao_get_shared_array_nwriters, taolib), Cint,
          (Ptr{Cvoid},), obj)

adjust_nwriters!(obj::SharedArray, adj::Integer) =
    ccall((:tao_adjust_shared_array_nwriters, taolib), Cint,
          (Ptr{Cvoid}, Cint), obj, adj)

get_counter(obj::SharedArray) =
    ccall((:tao_get_shared_array_counter, taolib), Int64,
          (Ptr{Cvoid},), obj)

set_counter!(obj::SharedArray, cnt::Integer) =
    ccall((:tao_set_shared_array_counter, taolib), Cvoid,
          (Ptr{Cvoid}, Int64), obj, cnt)

function get_timestamp(obj::SharedArray)
    ts_sec = Ref{Int64}()
    ts_nsec = Ref{Int64}()
    ccall((:tao_get_shared_array_timestamp, taolib), Cvoid,
          (Ptr{Cvoid}, Ptr{Int64}, Ptr{Int64}),
          obj, ts_sec, ts_nsec)
    return TimeStamp(ts_sec[], ts_nsec[])
end

set_timestamp!(obj::SharedArray, secs::Real) =
    set_timestamp!(obj, TimeStamp(secs))

set_timestamp!(obj::SharedArray, ts::TimeStamp) =
    ccall((:tao_set_shared_array_timestamp, taolib), Cvoid,
          (Ptr{Cvoid}, Int64, Int64), obj, ts.sec, ts.nsec)
